tsfc/coffee_mode.py

```diff
@@ -0,0 +1,317 @@
+                                one multiindex for each argumen
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+
+    def classify(argument_indices, expression):
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, (Indexed, Conditional)):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+
+    argument_indices = tuple([i for indices in argument_multiindices for i in indices])
+    # Apply argument factorisation unconditionally
+    classifier = partial(classify, set(argument_indices))
+    monomial_sums = collect_monomials(expressions, classifier)
+    monomial_sums = [optimise_monomial_sum(ms, argument_indices) for ms in monomial_sums]
+    return list(map(monomial_sum_to_expression, monomial_sums))
+
+
+def index_extent(factor, argument_indices):
+    """Compute the product of the extents of argument indices of a GEM expression
+
+    :arg factor: GEM expression
+    :arg argument_indices: set of argument indices
+
+    :returns: product of extents of argument indices
+    """
+    return numpy.product([i.extent for i in set(factor.free_indices).intersection(argument_indices)])
+
+
+def unique_sum_indices(monomial_sum):
+    """Create a generator of unique sum indices of monomials in a monomial sum.
+
+    :arg monomial_sum: :class:`MonomialSum` object
+
+    :returns: a generator of unique sum indices
+    """
+    seen = set()
+    for monomial in monomial_sum:
+        fs = frozenset(monomial.sum_indices)
+        if fs not in seen:
+            seen.add(fs)
+            yield monomial.sum_indices
+
+
+def monomial_sum_to_expression(monomial_sum):
+    """Convert a monomial sum to a GEM expression. Uses associate_product() and
+    associate_sum() to promote hoisting in the subsequent code generation.
+
+    :arg monomial_sum: :class:`MonomialSum` object
+
+    :returns: GEM expression
+    """
+    indexsums = []  # The result is summation of indexsums
+    sum_indices_set_map = {}  # fronzenset(sum_indices) -> sum_indices
+    monomial_groups = OrderedDict()  # frozonset(sum_indices) -> [(atomics, rest)]
+    # Group monomials according to their sum indices
+    for monomial in monomial_sum:
+        if not monomial.sum_indices:
+            # IndexSum(reduce(Product, atomics, rest), sum_indices)
+            product, _ = associate_product(monomial.atomics + (monomial.rest,))
+            indexsums.append(product)
+        else:
+            fs = frozenset(monomial.sum_indices)
+            sum_indices_set_map.setdefault(fs, monomial.sum_indices)
+            monomial_groups.setdefault(fs, []).append((monomial.atomics, monomial.rest))
```

Comment:
While you do call that new function at two different places, though in a rather useless manner. First, you call it at the beginning of `optimise_monomial_sum` which is legit. Then having optimised each subset separately, you merge them again with `MonomialSum.sum`. Only to split them again at the beginning of `monomial_sum_to_expression` (which always operates on the immediate return of `optimise_monomial_sum`). There is an optimisation/simplification opportunity!

Considering the broader picture, we have to go further than just giving names to blocks of code. You must learn about higher-order functions. We can do some practice tomorrow if you are in WPL.